src/api/rest/decorators/with-reducers.js

```diff
@@ -0,0 +1,128 @@
+import * as Actions from 'api/rest/actions';
+import { normalize } from 'normalizr';
+import { Alert } from 'react-native';
+
+const displayError = error => {
+  Alert.alert('API Error', error);
+};
+
+export const withReducers = Provider => {
+  const client = new Provider();
+
+  return new Proxy(withReducers, {
+    get: (c, namespace) => {
+      return new Proxy(client[namespace], {
+        get: (endpoint, call) => (...args) => (dispatch, getState) => {
+          // Used as a key for state.pagination
+          const actionName = `${namespace}_${call}`
+            .replace(/([A-Z])/g, '_$1')
+            .toUpperCase();
+
+          const action = Actions[actionName];
+
+          if (!action) {
+            return displayError(
+              `Unknown action. Did you forget to define Actions.${actionName}?`
+            );
+          }
+
+          // Identify if we have our magical last argument
+          const declaredArgsNumber = endpoint[call].length;
+          const isMagicArgAvailable = args.length === declaredArgsNumber;
+
+          const pureArgs = isMagicArgAvailable
+            ? args.slice(0, args.length - 1)
+            : args;
+          const magicArg = isMagicArgAvailable ? args[args.length - 1] : {};
+
+          const paginator = getState().pagination[actionName];
+
+          // pagination or entity ? <-
+
+          // Get accessToken from state
+          client.setAccessToken(getState().auth.accessToken);
+
+          let finalArgs = args;
+
+          if (paginator) {
+            const { loadMore = false } = magicArg;
+            const { pageCount = 0, isFetching = false, nextPageUrl } =
+              paginator[pureArgs.join('-')] || {};
```

Comment:
Why need `.join('-')`?

Reply:
example:

```js
client.activity.getNotifications(false, false); // unread
client.activity.getNotifications(true, false); // participating
```

We need to paginators here, and to identify them uniquely we use .join('-') to create the key: 

![screen shot 2017-10-15 at 6 38 12 pm](https://user-images.githubusercontent.com/304450/31587336-687935ea-b1e0-11e7-86e0-6070dc650046.png)

Reply:
Oh, I see, although it's not very clear.